fastlane/lib/fastlane/actions/github_api.rb

```diff
@@ -0,0 +1,266 @@
+module Fastlane
+  module Actions
+    module SharedValues
+      GITHUB_API_STATUS_CODE = :GITHUB_API_STATUS_CODE
+      GITHUB_API_RESPONSE = :GITHUB_API_RESPONSE
+      GITHUB_API_JSON = :GITHUB_API_JSON
+    end
+
+    class GithubApiAction < Action
+      def self.run(params)
+        require 'json'
+
+        http_method = (params[:http_method] || 'GET').to_s.upcase
+        url = parse_url(params[:server_url], params[:path], params[:url])
+        headers = parse_headers(params[:api_token], params[:headers])
+        request_body = parse_body(params[:body], params[:raw_body])
+        handled_errors = params[:errors] || {}
+
+        response = call_endpoint(
+          url,
+          http_method,
+          headers,
+          request_body,
+          { secure: params[:secure], debug: params[:debug] }
+        )
+
+        status_code = response[:status]
+        result = {
+          status: status_code,
+          body: response.body || "",
+          json: parse_json(response.body) || {}
+        }
+
+        if status_code.between?(200, 299)
+          if params[:debug]
+            UI.message("Response:")
+            UI.message(response.body)
+            UI.message("---")
+          end
+          yield(result) if block_given?
+        else
+          handled_error = handled_errors[status_code] || handled_errors['*']
+          if handled_error
+            handled_error.call(result)
+          else
+            UI.error("---")
+            UI.error("Request failed:\n#{http_method}: #{url}")
+            UI.error("Headers:\n#{headers}")
+            UI.error("---")
+            UI.error("Response:")
+            UI.error(response.body)
+            UI.user_error!("GitHub responded with #{status_code}\n---\n#{response.body}")
+          end
+        end
+
+        Actions.lane_context[SharedValues::GITHUB_API_STATUS_CODE] = result[:status]
+        Actions.lane_context[SharedValues::GITHUB_API_RESPONSE] = result[:body]
+        Actions.lane_context[SharedValues::GITHUB_API_JSON] = result[:json]
+
+        return result
+      end
+
+      def self.parse_headers(api_token, overrides)
+        require 'base64'
+        headers = { 'User-Agent' => 'fastlane-github_api' }
+        headers['Authorization'] = "Basic #{Base64.strict_encode64(api_token)}" if api_token
+        headers.merge(overrides || {})
+      end
+      private_class_method :parse_headers
```

Comment:
was following other examples in the code base, but I agree this would be much cleaner, especially in a file like this with so many private methods, wrapping them all up.